README
Data-Lock

$Id: README,v 1.0 2013/04/03 06:49:25 dankogai Exp dankogai $

This distribution contains Data::Lock and Attribute::Constant.

INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Data::Lock

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Data-Lock

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Data-Lock

    CPAN Ratings
        http://cpanratings.perl.org/d/Data-Lock

    Search CPAN
        http://search.cpan.org/dist/Data-Lock


COPYRIGHT AND LICENCE

Copyright (C) 2008-2013 Dan Kogai

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


Changes
# Revision history for Data-Lock
#
# $Id: Changes,v 1.2 2013/05/13 15:31:54 dankogai Exp dankogai $
#
$Revision: 1.2 $ $Date: 2013/05/13 15:31:54 $
! lib/Data/Lock.pm MANIFEST
+ t/05-dollar01.t
  Addressed: RT #93617
  https://rt.cpan.org/Public/Bug/Display.html?id=93617

1.02 2013/05/13 15:31:54
! t/00-load.t
  s/Attribute::Constant/Data::Lock/
! lib/Data/Lock.pm t/04-dlock.t
  Addressed: dunlock \%hash; # does not work
  https://rt.cpan.org/Public/Bug/Display.html?id=85233

1.01 2013/04/03 14:37:57 $
! lib/*/*.pm t/benchmark.t
  + Revised POD
  + updated benchmark

1.00 2013/04/03 06:49:25
! *
  Distro renamed from Attribute-Constant to Data-Lock.
  The use of Attribute is somewhat deprecated these days.

0.06 2013/03/20 22:17:40
! lib/Data/Lock.pm t/04-dlock.t
  Fixed: d(un)?lock vs. circular reference

0.05 2013/03/20 22:17:40
! lib/Data/Lock.pm t/04-dlock.t
  Addressed: dlock("0") locks $0
  dlock() and dunlock now simply ignores constants.
  https://rt.cpan.org/Public/Bug/Display.html?id=84092

0.04 2013/02/26 03:19:02
! t/03-obj.t
  Wrong Skip count was causing failure for perl 5.9 and below

0.03 2013/02/24 07:03:27
! *.pm t/03-obj.t
  You can now dlock() and dunlock() objects
  https://rt.cpan.org/Ticket/Display.html?id=62021

0.02 2008/06/27 19:50:52
! lib/Attribute/Constant.pm  t/03-obj.t
  Perl 5.8.x supported -- with caveats.  See POD

0.01 2008/06/27 19:11:42
+ * 
  First Release.



t/benchmark.pl
#!perl
#
# $Id: benchmark.pl,v 1.1 2013/04/03 14:37:57 dankogai Exp dankogai $
#
use strict;
use warnings;
use Benchmark qw/timethese cmpthese/;
use Data::Lock qw/dlock dunlock/;
use Attribute::Constant;
use Readonly;

dlock my $sd = 1;
my $sa : Constant(1);
Readonly my $sr => 1;
local *sg = \1; our $sg;

cmpthese(
    timethese(
        0,
        {
            dlock => sub {
                eval { $sd++ }; die unless $@; $sd == 1 or die;
            },
            Attribute => sub{
                eval { $sa++ }; die unless $@; $sa == 1 or die;
            },
            Readonly => sub {
                eval { $sr++ }; die unless $@; $sr == 1 or die;
            },
            glob => sub {
                eval { $sg++ }; die unless $@; $sg == 1 or die;
            },
        }
    )
);

dlock my $da = [ 1 .. 1000 ];
my @aa : Constant( 1 .. 1000 );
Readonly my @ar => ( 1 .. 1000 );

cmpthese(
    timethese(
        0,
        {
            dlock => sub{
		eval { pop @$da }; die unless $@; $da->[0] == 1 or die;
	    },
            Attribute => sub{
		eval { pop @aa }; die unless $@; $aa[0] == 1 or die;
	    },
            Readonly => sub {
		eval { pop @ar }; die unless $@; $ar[0] == 1 or die;
	    },
	}
    )
);

dlock my $dh = { map { $_ => $_*$_ } 1 .. 1000 }; 
my %ha : Constant( map { $_ => $_*$_ } 1 .. 1000 );
Readonly my %hr => ( map { $_ => $_ * $_ } 1 .. 1000 );

cmpthese(
    timethese(
        0,
        {
            dlock => sub{
		eval { $dh->{zero}++ }; die unless $@; $dh->{1000} == 1e6 or die;
	    },
	    Attribute => sub{
		eval { $ha{zero}++ }; die unless $@; $ha{1000} == 1e6 or die;
	    },
            Readonly => sub {
		eval { $hr{zero}++ }; die unless $@; $hr{1000} == 1e6 or die;
	    },
        }
    )
);


lib/Attribute/Constant.pm
package Attribute::Constant;
use 5.008001;
use warnings;
use strict;
our $VERSION = sprintf "%d.%02d", q$Revision: 1.1 $ =~ /(\d+)/g;
use Attribute::Handlers;
use Data::Lock ();

sub UNIVERSAL::Constant : ATTR {
    my ( $pkg, $sym, $ref, $attr, $data, $phase ) = @_;
    (
          ref $ref eq 'HASH'  ? %$ref
        : ref $ref eq 'ARRAY' ? @$ref
        :                       ($$ref)
      )
      = ref $data
      ? ref $data eq 'ARRAY'
          ? @$data    # perl 5.10.x
          : $data
      : $data;        # perl 5.8.x
    Data::Lock::dlock($ref);
}

1;
__END__

=head1 NAME

Attribute::Constant - Make read-only variables via attribute

=head1 VERSION

$Id: Constant.pm,v 1.1 2013/04/03 14:37:57 dankogai Exp dankogai $

=head1 SYNOPSIS

 use Attribute::Constant;
 my $sv : Constant( $initial_value );
 my @av : Constant( @values );
 my %hv : Constant( key => value, key => value, ...);

=head1 DESCRIPTION

This module uses L<Data::Lock> to make the variable read-only.  Check
the document and source of L<Data::Lock> for its mechanism.

=head1 ATTRIBUTES

This module adds only one attribute, C<Constant>.  You give its
initial value as shown.  Unlike L<Readonly>, parantheses cannot be
ommited but it is semantically more elegant and thanks to
L<Data::Lock>, it imposes almost no performance penalty.

=head1 CAVEAT

=head2 Multi-line attributes

Multi-line attributes are not allowed in Perl 5.8.x.

  my $o : Constant(Foo->new(one=>1,two=>2,three=>3));    # ok
  my $p : Constant(Bar->new(
                            one   =>1,
                            two   =>2,
                            three =>3
                           )
                 ); # needs Perl 5.10

In which case you can use L<Data::Lock> instead:

  dlock(my $p = Bar->new(
        one   => 1,
        two   => 2,
        three => 3
    )
  );

After all, this module is a wrapper to L<Data::Lock>;

=head2 Constants from Variables

You may be surprised the following code B<DOES NOT> work as you expected:

  #!/usr/bin/perl
  use strict;
  use warnings;
  use Attribute::Constant;
  use Data::Dumper;
  {
    package MyClass;
    sub new {
        my ( $class, %params ) = @_;
        return bless \%params, $class;
    }
  }
  my $o = MyClass->new( a => 1, b => 2 );
  my $x : Constant($o);
  print Dumper( $o, $x );

Which outputs:

  $VAR1 = bless( {
                 'a' => 1,
                 'b' => 2
               }, 'MyClass' );
  $VAR2 = undef;

Why?  Because C< $x : Constant($o) > happens B<before>
C<< $o = Myclass->new() >>.

On the other hand, the following works.

  my $y : Constant(MyClass->new(a => 1,b => 2));
  print Dumper( $o, $y );

Rule of the thumb is do not feed variables to constant because
varialbes change after the attribute invocation.

Or simply use C<Data::Lock::dlock>.

  use Data::Lock qw/dlock/;
  dlock my $z = $o;
  print Dumper( $o, $y );

=head1 SEE ALSO

L<Data::Lock>, L<constant>

=head1 AUTHOR

Dan Kogai, C<< <dankogai+cpan at gmail.com> >>

=head1 BUGS & SUPPORT

See L<Data::Lock>.

=head1 ACKNOWLEDGEMENTS

L<Readonly>

=head1 COPYRIGHT & LICENSE

Copyright 2008-2013 Dan Kogai, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


lib/Data/Lock.pm
package Data::Lock;
use 5.008001;
use warnings;
use strict;
our $VERSION = sprintf "%d.%02d", q$Revision: 1.2 $ =~ /(\d+)/g;

use Attribute::Handlers;
use Scalar::Util ();

use base 'Exporter';
our @EXPORT_OK = qw/dlock dunlock/;

#my @builtin_types = 
#    qw/SCALAR ARRAY HASH CODE REF GLOB LVALUE FORMAT IO VSTRING Regexp/;

for my $locked ( 0, 1 ) {
    my $subname = $locked ? 'dlock' : 'dunlock';
    no strict 'refs';
    *{$subname} = sub {
        no warnings "uninitialized";
        return if $_[1] and Internals::SvREADONLY( $_[0]) == $locked;
        Internals::SvREADONLY( $_[0], $locked );
        return unless my $type = Scalar::Util::reftype( $_[0] );
        for (
              $type eq 'ARRAY' ? @{ $_[0] }
            : $type eq 'HASH'  ? values %{ $_[0] }
            : $type ne 'CODE'  ? ${ $_[0] }
            :                    ()
          )
        {
            &$subname($_, 1) if ref $_;
            Internals::SvREADONLY( $_, $locked );
        }
            $type eq 'ARRAY' ? Internals::SvREADONLY( @{ $_[0] }, $locked )
          : $type eq 'HASH'  ? Internals::SvREADONLY( %{ $_[0] }, $locked )
          : $type ne 'CODE'  ? Internals::SvREADONLY( ${ $_[0] }, $locked )
          :                    undef;
    };
}

1;
__END__

=head1 NAME

Data::Lock - makes variables (im)?mutable

=head1 VERSION

$Id: Lock.pm,v 1.2 2013/05/13 15:31:54 dankogai Exp dankogai $

=head1 SYNOPSIS

   use Data::Lock qw/dlock dunlock/;

   dlock my $sv = $initial_value;
   dlock my $ar = [@values];
   dlock my $hr = { key => value, key => value, ... };
   dunlock $sv;
   dunlock $ar; dunlock \@av;
   dunlock $hr; dunlock \%hv;

=head1 DESCRIPTION

C<dlock> makes the specified variable immutable like L<Readonly>.
Unlike L<Readonly> which implements immutability via C<tie>, C<dlock>
makes use of the internal flag of perl SV so it imposes almost no
penalty.

Like L<Readonly>, C<dlock> locks not only the variable itself but also
elements therein.

As of verion 0.03, you can C<dlock> objects as well.  Below is an
example constructor that returns an immutable object:

  sub new {
      my $pkg = shift;
      my $self = { @_ };
      bless $self, $pkg;
      dlock($self);
      $self;
  }

Or consider using L<Moose>.

=head1 EXPORT

Like L<List::Util> and L<Scalar::Util>, functions are exported only
explicitly. This module comes with C<dlock> and C<dunlock>.

  use Data::Lock;                   # nothing imported;
  use Data::Lock qw/dlock dunlock/; # imports dlock() and dunlock()

=head1 FUNCTIONS

=head2 dlock

  dlock($scalar);

Locks $scalar and if $scalar is a reference, recursively locks referents.

=head2 dunlock

Does the opposite of C<dlock>.

=head1 BENCHMARK

Here I have benchmarked like this.

  1.  Create an immutable variable.
  2.  try to change it and see if it raises exception
  3.  make sure the value stored remains unchanged.

See F<t/benchmark.pl> for details.

=over 2

=item Simple scalar

                Rate  Readonly Attribute      glob     dlock
  Readonly   11987/s        --      -98%      -98%      -98%
  Attribute 484562/s     3943%        --       -1%       -4%
  glob      487239/s     3965%        1%        --       -3%
  dlock     504247/s     4107%        4%        3%        --

=item Array with 1000 entries

                Rate  Readonly     dlock Attribute
  Readonly   12396/s        --      -97%      -97%
  dlock     444703/s     3488%        --       -6%
  Attribute 475557/s     3736%        7%        --

=item Hash with 1000 key/value pairs

                Rate  Readonly     dlock Attribute
  Readonly   10855/s        --      -97%      -97%
  dlock     358867/s     3206%        --       -5%
  Attribute 377087/s     3374%        5%        --

=back

=head1 SEE ALSO

L<Readonly>, L<perlguts>, L<perlapi>

=head1 AUTHOR

Dan Kogai, C<< <dankogai+gmail at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-data-lock at
rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Data-Lock>.  I will
be notified, and then you'll automatically be notified of progress on
your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Data::Lock

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Data-Lock>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Data-Lock>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Data-Lock>

=item * Search CPAN

L<http://search.cpan.org/dist/Data-Lock>

=back

=head1 COPYRIGHT & LICENSE

Copyright 2008-2013 Dan Kogai, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.
